File: framework/languages/node/e2e/PubSub/Server/Publisher/Infrastructure/evidence-store.ts
import fs from 'node:fs';
import path from 'node:path';

export class EvidenceStore {
  readonly rid: string;
  private readonly entries: string[] = [];

  constructor(private readonly filePath?: string) {
    this.rid = process.env.ZLINK_E2E_RID ?? 'publisher';
    if (filePath !== undefined && filePath.length > 0) {
      fs.mkdirSync(path.dirname(filePath), { recursive: true });
      fs.writeFileSync(filePath, '');
    }
  }

  add(entry: string): void {
    this.entries.push(entry);
    if (this.filePath !== undefined && this.filePath.length > 0) {
      fs.appendFileSync(this.filePath, `${entry}\n`);
    }
  }

  snapshot(): readonly string[] {
    return [...this.entries];
  }

  clear(): void {
    this.entries.length = 0;
    if (this.filePath !== undefined && this.filePath.length > 0) {
      fs.writeFileSync(this.filePath, '');
    }
  }
}

File: framework/languages/dotnet/e2e/LocationMessaging/Server/Workflow/Infrastructure/EvidenceStore.cs
using System.Collections.Concurrent;

namespace LocationMessaging.Server.Workflow.Infrastructure;

internal sealed class EvidenceStore
{
    private readonly ConcurrentQueue<string> _entries = new();
    private readonly object _fileGate = new();
    private readonly string? _filePath;
    private readonly SemaphoreSlim _signal = new(0);

    public EvidenceStore(string rid, string? filePath)
    {
        _filePath = filePath;
        Rid = rid;
        if (!string.IsNullOrWhiteSpace(_filePath))
        {
            Directory.CreateDirectory(Path.GetDirectoryName(_filePath)!);
            File.WriteAllText(_filePath, string.Empty);
        }
    }

    public string Rid { get; }

    public void Add(string entry)
    {
        _entries.Enqueue(entry);
        _signal.Release();
        if (string.IsNullOrWhiteSpace(_filePath))
        {
            return;
        }

        lock (_fileGate)
        {
            File.AppendAllText(_filePath, entry + Environment.NewLine);
        }
    }

    public string[] Snapshot() => _entries.ToArray();

    public async Task<string[]> WaitUntilAsync(
        Func<string, bool> predicate,
        TimeSpan timeout,
        CancellationToken cancellationToken)
    {
        var deadline = DateTimeOffset.UtcNow + timeout;
        while (true)
        {
            var snapshot = Snapshot();
            if (snapshot.Any(predicate))
            {
                return snapshot;
            }

            var remaining = deadline - DateTimeOffset.UtcNow;
            if (remaining <= TimeSpan.Zero
                || !await _signal.WaitAsync(remaining, cancellationToken))
            {
                return Snapshot();
            }
        }
    }

    public void Clear()
    {
        while (_entries.TryDequeue(out _))
        {
        }

        if (!string.IsNullOrWhiteSpace(_filePath))
        {
            lock (_fileGate)
            {
                File.WriteAllText(_filePath, string.Empty);
            }
        }
    }
}
